StackOverflowError with generic types

There's a circle in `TypeName.get/list`, `ParameterizedTypeName.get` and `TypeVariableName.get`. The following test:
```java
  private <E extends Enum<E>> E recursiveEnum(E[] values) {
    return values[0];
  }
  
  @Test public void genericTypes() throws Exception {
    Method recursiveEnum = getClass().getDeclaredMethod("recursiveEnum", Enum[].class);
    TypeName.get(recursiveEnum.getReturnType());
    TypeName.get(recursiveEnum.getGenericReturnType()); // StackOverflowError !!!
    TypeName.get(recursiveEnum.getParameterTypes()[0]);
    TypeName.get(recursiveEnum.getGenericParameterTypes()[0]); // StackOverflowError !!!
  }
```
produces a endless stack trace like:
```
Caused by: java.lang.StackOverflowError
	at java.lang.Class.getSimpleBinaryName(Class.java:1443)
	at java.lang.Class.getSimpleName(Class.java:1309)
	at com.squareup.javapoet.ClassName.get(ClassName.java:113)
	at com.squareup.javapoet.ParameterizedTypeName.get(ParameterizedTypeName.java:76)
	at com.squareup.javapoet.TypeName.get(TypeName.java:260)
	at com.squareup.javapoet.TypeName.list(TypeName.java:280)
	at com.squareup.javapoet.TypeVariableName.get(TypeVariableName.java:129)
	at com.squareup.javapoet.TypeName.get(TypeName.java:266)
	at com.squareup.javapoet.TypeName.list(TypeName.java:280)
	at com.squareup.javapoet.ParameterizedTypeName.get(ParameterizedTypeName.java:77)
	at com.squareup.javapoet.TypeName.get(TypeName.java:260)
	at com.squareup.javapoet.TypeName.list(TypeName.java:280)
	at com.squareup.javapoet.TypeVariableName.get(TypeVariableName.java:129)
	at com.squareup.javapoet.TypeName.get(TypeName.java:266)
	at com.squareup.javapoet.TypeName.list(TypeName.java:280)
	at com.squareup.javapoet.ParameterizedTypeName.get(ParameterizedTypeName.java:77)
	at com.squareup.javapoet.TypeName.get(TypeName.java:260)
```

--- a/src/main/java/com/squareup/javapoet/ArrayTypeName.java
+++ b/src/main/java/com/squareup/javapoet/ArrayTypeName.java
@@ -84,6 +84,10 @@ public final class ArrayTypeName extends TypeName {
 
   /** Returns an array type equivalent to {@code type}. */
   public static ArrayTypeName get(GenericArrayType type) {
-    return ArrayTypeName.of(get(type.getGenericComponentType()));
+    return get(type, new LinkedHashMap<Type, TypeVariableName>());
+  }
+
+  static ArrayTypeName get(GenericArrayType type, Map<Type, TypeVariableName> map) {
+    return ArrayTypeName.of(get(type.getGenericComponentType(), map));
   }
 }

--- a/src/main/java/com/squareup/javapoet/ParameterizedTypeName.java
+++ b/src/main/java/com/squareup/javapoet/ParameterizedTypeName.java
@@ -20,7 +20,9 @@ import java.lang.reflect.ParameterizedType;
 import java.lang.reflect.Type;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.LinkedHashMap;
 import java.util.List;
+import java.util.Map;
 
 import static com.squareup.javapoet.Util.checkArgument;
 import static com.squareup.javapoet.Util.checkNotNull;
@@ -89,7 +91,12 @@ public final class ParameterizedTypeName extends TypeName {
 
   /** Returns a parameterized type equivalent to {@code type}. */
   public static ParameterizedTypeName get(ParameterizedType type) {
+    return get(type, new LinkedHashMap<Type, TypeVariableName>());
+  }
+
+  /** Returns a parameterized type equivalent to {@code type}. */
+  static ParameterizedTypeName get(ParameterizedType type, Map<Type, TypeVariableName> map) {
     return new ParameterizedTypeName(ClassName.get((Class<?>) type.getRawType()),
-        TypeName.list(type.getActualTypeArguments()));
+        TypeName.list(type.getActualTypeArguments(), map));
   }
 }

--- a/src/main/java/com/squareup/javapoet/TypeName.java
+++ b/src/main/java/com/squareup/javapoet/TypeName.java
@@ -261,6 +261,10 @@ public class TypeName {
 
   /** Returns a type name equivalent to {@code type}. */
   public static TypeName get(Type type) {
+    return get(type, new LinkedHashMap<Type, TypeVariableName>());
+  }
+
+  static TypeName get(Type type, Map<Type, TypeVariableName> map) {
     if (type instanceof Class<?>) {
       Class<?> classType = (Class<?>) type;
       if (type == void.class) return VOID;
@@ -272,20 +276,20 @@ public class TypeName {
       if (type == char.class) return CHAR;
       if (type == float.class) return FLOAT;
       if (type == double.class) return DOUBLE;
-      if (classType.isArray()) return ArrayTypeName.of(get(classType.getComponentType()));
+      if (classType.isArray()) return ArrayTypeName.of(get(classType.getComponentType(), map));
       return ClassName.get(classType);
 
     } else if (type instanceof ParameterizedType) {
-      return ParameterizedTypeName.get((ParameterizedType) type);
+      return ParameterizedTypeName.get((ParameterizedType) type, map);
 
     } else if (type instanceof WildcardType) {
-      return WildcardTypeName.get((WildcardType) type);
+      return WildcardTypeName.get((WildcardType) type, map);
 
     } else if (type instanceof TypeVariable<?>) {
-      return TypeVariableName.get((TypeVariable<?>) type);
+      return TypeVariableName.get((TypeVariable<?>) type, map);
 
     } else if (type instanceof GenericArrayType) {
-      return ArrayTypeName.get((GenericArrayType) type);
+      return ArrayTypeName.get((GenericArrayType) type, map);
 
     } else {
       throw new IllegalArgumentException("unexpected type: " + type);
@@ -294,9 +298,13 @@ public class TypeName {
 
   /** Converts an array of types to a list of type names. */
   static List<TypeName> list(Type[] types) {
+    return list(types, new LinkedHashMap<Type, TypeVariableName>());
+  }
+
+  static List<TypeName> list(Type[] types, Map<Type, TypeVariableName> map) {
     List<TypeName> result = new ArrayList<>(types.length);
     for (Type type : types) {
-      result.add(get(type));
+      result.add(get(type, map));
     }
     return result;
   }

--- a/src/main/java/com/squareup/javapoet/TypeVariableName.java
+++ b/src/main/java/com/squareup/javapoet/TypeVariableName.java
@@ -20,6 +20,7 @@ import java.lang.reflect.Type;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
+import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 
@@ -139,6 +140,23 @@ public final class TypeVariableName extends TypeName {
 
   /** Returns type variable equivalent to {@code type}. */
   public static TypeVariableName get(java.lang.reflect.TypeVariable<?> type) {
-    return TypeVariableName.of(type.getName(), TypeName.list(type.getBounds()));
+    return get(type, new LinkedHashMap<Type, TypeVariableName>());
+  }
+
+  /** @see #get(java.lang.reflect.TypeVariable, Map) */
+  static TypeVariableName get(java.lang.reflect.TypeVariable<?> type,
+      Map<Type, TypeVariableName> map) {
+    TypeVariableName result = map.get(type);
+    if (result == null) {
+      List<TypeName> bounds = new ArrayList<>();
+      List<TypeName> visibleBounds = Collections.unmodifiableList(bounds);
+      result = new TypeVariableName(type.getName(), visibleBounds);
+      map.put(type, result);
+      for (Type bound : type.getBounds()) {
+        bounds.add(TypeName.get(bound, map));
+      }
+      bounds.remove(OBJECT);
+    }
+    return result;
   }
 }

--- a/src/main/java/com/squareup/javapoet/WildcardTypeName.java
+++ b/src/main/java/com/squareup/javapoet/WildcardTypeName.java
@@ -129,8 +129,12 @@ public final class WildcardTypeName extends TypeName {
   }
 
   public static TypeName get(WildcardType wildcardName) {
+    return get(wildcardName, new LinkedHashMap<Type, TypeVariableName>());
+  }
+
+  static TypeName get(WildcardType wildcardName, Map<Type, TypeVariableName> map) {
     return new WildcardTypeName(
-        list(wildcardName.getUpperBounds()),
-        list(wildcardName.getLowerBounds()));
+        list(wildcardName.getUpperBounds(), map),
+        list(wildcardName.getLowerBounds(), map));
   }
 }

--- a/src/test/java/com/squareup/javapoet/TypeNameTest.java
+++ b/src/test/java/com/squareup/javapoet/TypeNameTest.java
@@ -16,61 +16,85 @@
 package com.squareup.javapoet;
 
 import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotEquals;
 
 import java.io.Serializable;
+import java.lang.reflect.Method;
 import java.rmi.server.UID;
 import java.util.Comparator;
+import java.util.List;
 import java.util.Set;
-
-import org.junit.Assert;
 import org.junit.Test;
 
 public class TypeNameTest {
 
+  protected <E extends Enum<E>> E generic(E[] values) {
+    return values[0];
+  }
+
+  @Test public void genericType() throws Exception {
+    Method recursiveEnum = getClass().getDeclaredMethod("generic", Enum[].class);
+    TypeName.get(recursiveEnum.getReturnType());
+    TypeName.get(recursiveEnum.getGenericReturnType());
+    TypeName.get(recursiveEnum.getParameterTypes()[0]);
+    TypeName.get(recursiveEnum.getGenericParameterTypes()[0]);
+  }
+
   @Test public void equalsAndHashCodePrimitive() {
-    test(TypeName.BOOLEAN, TypeName.BOOLEAN);
-    test(TypeName.BYTE, TypeName.BYTE);
-    test(TypeName.CHAR, TypeName.CHAR);
-    test(TypeName.DOUBLE, TypeName.DOUBLE);
-    test(TypeName.FLOAT, TypeName.FLOAT);
-    test(TypeName.INT, TypeName.INT);
-    test(TypeName.LONG, TypeName.LONG);
-    test(TypeName.SHORT, TypeName.SHORT);
-    test(TypeName.VOID, TypeName.VOID);
+    assertEqualsHashCodeAndToString(TypeName.BOOLEAN, TypeName.BOOLEAN);
+    assertEqualsHashCodeAndToString(TypeName.BYTE, TypeName.BYTE);
+    assertEqualsHashCodeAndToString(TypeName.CHAR, TypeName.CHAR);
+    assertEqualsHashCodeAndToString(TypeName.DOUBLE, TypeName.DOUBLE);
+    assertEqualsHashCodeAndToString(TypeName.FLOAT, TypeName.FLOAT);
+    assertEqualsHashCodeAndToString(TypeName.INT, TypeName.INT);
+    assertEqualsHashCodeAndToString(TypeName.LONG, TypeName.LONG);
+    assertEqualsHashCodeAndToString(TypeName.SHORT, TypeName.SHORT);
+    assertEqualsHashCodeAndToString(TypeName.VOID, TypeName.VOID);
   }
 
   @Test public void equalsAndHashCodeArrayTypeName() {
-    test(ArrayTypeName.of(Object.class), ArrayTypeName.of(Object.class));
-    test(TypeName.get(Object[].class), ArrayTypeName.of(Object.class));
-    // ? check(ClassName.bestGuess("java.lang.Object[]"), ArrayTypeName.of(Object.class));
+    assertEqualsHashCodeAndToString(ArrayTypeName.of(Object.class),
+        ArrayTypeName.of(Object.class));
+    assertEqualsHashCodeAndToString(TypeName.get(Object[].class),
+        ArrayTypeName.of(Object.class));
   }
 
   @Test public void equalsAndHashCodeClassName() {
-    test(ClassName.get(Object.class), ClassName.get(Object.class));
-    test(TypeName.get(Object.class), ClassName.get(Object.class));
-    test(ClassName.bestGuess("java.lang.Object"), ClassName.get(Object.class));
+    assertEqualsHashCodeAndToString(ClassName.get(Object.class), ClassName.get(Object.class));
+    assertEqualsHashCodeAndToString(TypeName.get(Object.class), ClassName.get(Object.class));
+    assertEqualsHashCodeAndToString(ClassName.bestGuess("java.lang.Object"),
+        ClassName.get(Object.class));
   }
-  
+
   @Test public void equalsAndHashCodeParameterizedTypeName() {
-    test(ParameterizedTypeName.get(Object.class), ParameterizedTypeName.get(Object.class));
-    test(ParameterizedTypeName.get(Set.class, UID.class), ParameterizedTypeName.get(Set.class, UID.class));
+    assertEqualsHashCodeAndToString(ParameterizedTypeName.get(Object.class),
+        ParameterizedTypeName.get(Object.class));
+    assertEqualsHashCodeAndToString(ParameterizedTypeName.get(Set.class, UID.class),
+        ParameterizedTypeName.get(Set.class, UID.class));
+    assertNotEquals(ClassName.get(List.class), ParameterizedTypeName.get(List.class,
+        String.class));
   }
-  
+
   @Test public void equalsAndHashCodeTypeVariableName() {
-    test(TypeVariableName.get(Object.class), TypeVariableName.get(Object.class));
-    TypeVariableName typeVariable1 = TypeVariableName.get("T", Comparator.class, Serializable.class);
-    TypeVariableName typeVariable2 = TypeVariableName.get("T", Comparator.class, Serializable.class);
-    test(typeVariable1, typeVariable2);
+    assertEqualsHashCodeAndToString(TypeVariableName.get(Object.class),
+        TypeVariableName.get(Object.class));
+    TypeVariableName typeVar1 = TypeVariableName.get("T", Comparator.class, Serializable.class);
+    TypeVariableName typeVar2 = TypeVariableName.get("T", Comparator.class, Serializable.class);
+    assertEqualsHashCodeAndToString(typeVar1, typeVar2);
   }
 
   @Test public void equalsAndHashCodeWildcardTypeName() {
-    test(WildcardTypeName.subtypeOf(Object.class), WildcardTypeName.subtypeOf(Object.class));
-    test(WildcardTypeName.subtypeOf(Serializable.class), WildcardTypeName.subtypeOf(Serializable.class));
-    test(WildcardTypeName.supertypeOf(String.class), WildcardTypeName.supertypeOf(String.class));
+    assertEqualsHashCodeAndToString(WildcardTypeName.subtypeOf(Object.class),
+        WildcardTypeName.subtypeOf(Object.class));
+    assertEqualsHashCodeAndToString(WildcardTypeName.subtypeOf(Serializable.class),
+        WildcardTypeName.subtypeOf(Serializable.class));
+    assertEqualsHashCodeAndToString(WildcardTypeName.supertypeOf(String.class),
+        WildcardTypeName.supertypeOf(String.class));
   }
 
-  private void test(TypeName a, TypeName b) {
-    Assert.assertEquals(a.toString(), b.toString());
+  private void assertEqualsHashCodeAndToString(TypeName a, TypeName b) {
+    assertEquals(a.toString(), b.toString());
     assertThat(a.equals(b)).isTrue();
     assertThat(a.hashCode()).isEqualTo(b.hashCode());
   }

